# Imports

In [1]:
import os
import pandas as pd
from huggingface_hub import login

import timm
from PIL import Image

import torch
from torchvision import transforms
from gigapath.pipeline import tile_one_slide, load_tile_slide_encoder, run_inference_with_tile_encoder, run_inference_with_slide_encoder

torch.cuda.empty_cache()

/home/ge26xaj/miniconda3/envs/gigapath/lib/python3.9/site-packages/timm/models/registry.py:4: FutureWarning: Importing from timm.models.registry is deprecated, please import via timm.models
  warnings.warn(f"Importing from {__name__} is deprecated, please import via timm.models", FutureWarning)
/home/ge26xaj/miniconda3/envs/gigapath/lib/python3.9/site-packages/timm/models/layers/__init__.py:48: FutureWarning: Importing from timm.models.layers is deprecated, please import via timm.layers
  warnings.warn(f"Importing from {__name__} is deprecated, please import via timm.layers", FutureWarning)


# Login to Huggingface

In [3]:
login("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")
!export HF_TOKEN="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"

Token will not been saved to git credential helper. Pass `add_to_git_credential=True` if you want to set the git credential as well.
Token is valid (permission: read).
Your token has been saved to /home/ge26xaj/.cache/huggingface/token
Login successful


# Create Encoders

In [4]:
# Older versions of timm have compatibility issues. Please ensure that you use a newer version by running the following command: pip install timm>=1.0.3.
tile_encoder = timm.create_model("hf_hub:prov-gigapath/prov-gigapath", pretrained=True)

transform = transforms.Compose(
    [
        transforms.Resize(256, interpolation=transforms.InterpolationMode.BICUBIC),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        transforms.Normalize(mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225)),
    ]
)

/home/ge26xaj/miniconda3/envs/gigapath/lib/python3.9/site-packages/torch/_utils.py:776: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  return self.fget.__get__(instance, owner)()


In [5]:
# Load the tile and slide encoder models
# NOTE: The CLS token is not trained during the slide-level pretraining.
# Here, we enable the use of global pooling for the output embeddings.
tile_encoder, slide_encoder_model = load_tile_slide_encoder(global_pool=True)

Tile encoder param # 1134953984
dilated_ratio:  [1, 2, 4, 8, 16]
segment_length:  [1024, 5792, 32768, 185363, 1048576]
Number of trainable LongNet parameters:  85148160
Global Pooling: True


slide_encoder.pth:   0%|          | 0.00/345M [00:00<?, ?B/s]

 Successfully Loaded Pretrained GigaPath model from hf_hub:prov-gigapath/prov-gigapath 
Slide encoder param # 86330880


# Create Tiles from Slide

In [ ]:
# assert "HF_TOKEN" in os.environ, "Please set the HF_TOKEN environment variable to your Hugging Face API token"
img_path = "../../../../../../../../mnt/data/Dataset/CAMELYON16/images/normal_041.tif"

# local_dir = os.path.join(os.path.expanduser("~"), ".cache/")
# huggingface_hub.hf_hub_download("prov-gigapath/prov-gigapath", filename="sample_data/PROV-000-000001.ndpi", local_dir=local_dir, force_download=True)
slide_path = os.path.join(img_path)

save_dir = os.path.join('outputs/preprocessing/')

print("NOTE: Prov-GigaPath is trained with 0.5 mpp preprocessed slides. Please make sure to use the appropriate level for the 0.5 MPP")
tile_one_slide(slide_path, save_dir=save_dir, level=1)

print("NOTE: tiling dependency libraries can be tricky to set up. Please double check the generated tile images.")

In [11]:
# load image tiles
slide_dir = "outputs/preprocessing/output/" + os.path.basename(slide_path) + "/"
image_paths = [os.path.join(slide_dir, img) for img in os.listdir(slide_dir) if img.endswith('.png')]

print(f"Found {len(image_paths)} image tiles")

Found 7258 image tiles


# Encode Tiles

In [6]:
tile_path = "./outputs/preprocessing/output/normal_040.tif/20480x_124928y.png"

sample_input = transform(Image.open(tile_path).convert("RGB")).unsqueeze(0)

tile_encoder.eval()
with torch.no_grad():
    output = tile_encoder(sample_input).squeeze()

In [7]:
torch.cuda.current_device()
torch.cuda.get_device_name(0)

'NVIDIA GeForce RTX 3090'

In [12]:
tile_encoder_outputs = run_inference_with_tile_encoder(image_paths, tile_encoder)

for k in tile_encoder_outputs.keys():
    print(f"tile_encoder_outputs[{k}].shape: {tile_encoder_outputs[k].shape}")

Running inference with tile encoder: 100%|██████████| 57/57 [01:25<00:00,  1.50s/it]

tile_encoder_outputs[tile_embeds].shape: torch.Size([7258, 1536])
tile_encoder_outputs[coords].shape: torch.Size([7258, 2])


# Encode Slides

In [13]:
# run inference with the slide encoder
slide_embeds = run_inference_with_slide_encoder(slide_encoder_model=slide_encoder_model, **tile_encoder_outputs)
print(slide_embeds.keys())

dict_keys(['layer_0_embed', 'layer_1_embed', 'layer_2_embed', 'layer_3_embed', 'layer_4_embed', 'layer_5_embed', 'layer_6_embed', 'layer_7_embed', 'layer_8_embed', 'layer_9_embed', 'layer_10_embed', 'layer_11_embed', 'layer_12_embed', 'last_layer_embed'])


In [14]:
print((slide_embeds["last_layer_embed"]).shape)

torch.Size([1, 768])


In [2]:
import pandas as pd

df = pd.read_excel("../../../../../../../../mnt/nfs03-R6/BRACS/BRACS.xlsx")
slides = df['WSI Filename']
len(slides)
df

,WSI Filename,Patient Id,RoI,WSI label,Set
0,BRACS_264,85,24,PB,Testing
1,BRACS_265,87,18,UDH,Validation
2,BRACS_280,30,19,IC,Training
3,BRACS_281,111,37,IC,Training
4,BRACS_283,109,49,IC,Training
...,...,...,...,...,...
542,BRACS_1003728,71,1,ADH,Training
543,BRACS_1003730,176,1,PB,Training
544,BRACS_1003731,153,1,IC,Training
545,BRACS_1003732,153,1,IC,Training


In [3]:
df[df["WSI Filename"]=="BRACS_1879"]

,WSI Filename,Patient Id,RoI,WSI label,Set
335,BRACS_1879,129,5,PB,Training


In [6]:
import os
from collections import defaultdict

# Specify the top-level directory
top_dir = "../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI//"

# Dictionary to store the counts for each folder
folder_counts = defaultdict(int)

# List to store all .svs file paths
svs_files = []

# Traverse the directory tree
for root, dirs, files in os.walk(top_dir):
    # Filter .svs files
    svs_in_folder = [file for file in files if file.lower().endswith(".svs")]
    
    # Count the .svs files in the current folder
    folder_counts[root] += len(svs_in_folder)
    
    # Append the full paths of the .svs files
    svs_files.extend([os.path.join(root, file) for file in svs_in_folder])

# Print the count of .svs files in each folder
print(len(svs_files))
print("\n.svs file counts per folder:")
for folder, count in folder_counts.items():
    if count > 0:
        print(f"{folder}: {count} files")

file_names = [os.path.basename(str(item)[:-4]) for item in svs_files]
# x =svs_files[0].split("/")[-1] #190
# x_split = x.split("_")
# x = f"{x_split[-2]}_{x_split[-1][:-3]}"
# df = pd.read_excel("../../../../../../mnt/nfs03-R6/BRACS/BRACS.xlsx")
# print(x)
# df[df["WSI Filename"]==x]
#547

546

.svs file counts per folder:
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI//train/Group_MT/Type_DCIS: 40 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI//train/Group_MT/Type_IC: 100 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI//train/Group_AT/Type_FEA: 24 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI//train/Group_AT/Type_ADH: 28 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI//train/Group_BT/Type_N: 27 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI//train/Group_BT/Type_PB: 120 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI//train/Group_BT/Type_UDH: 55 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI//val/Group_MT/Type_DCIS: 9 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI//val/Group_MT/Type_IC: 12 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI//val/Group_AT/Type_FEA: 6 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI//val/Group_AT/Type_ADH: 8 files
../../../.

In [23]:
pd.read_csv("../../fm_histopathology/scripts/CLAM/presets/tcga.csv")

,seg_level,sthresh,mthresh,close,use_otsu,a_t,a_h,max_n_holes,vis_level,line_thickness,white_thresh,black_thresh,use_padding,contour_fn,keep_ids,exclude_ids
0,-1,8,7,4,False,16,4,8,-1,100,5,50,True,four_pt,none,none


In [14]:
missing = []
for slide in slides:
    if slide not in file_names:
        missing.append(slide)

print(len(missing))
print(missing)

547
['BRACS_264', 'BRACS_265', 'BRACS_280', 'BRACS_281', 'BRACS_283', 'BRACS_286', 'BRACS_288', 'BRACS_291', 'BRACS_292', 'BRACS_295', 'BRACS_297', 'BRACS_298', 'BRACS_300', 'BRACS_301', 'BRACS_305', 'BRACS_310', 'BRACS_311', 'BRACS_734', 'BRACS_735', 'BRACS_736', 'BRACS_737', 'BRACS_738', 'BRACS_739', 'BRACS_740', 'BRACS_743', 'BRACS_745', 'BRACS_747', 'BRACS_748', 'BRACS_749', 'BRACS_750', 'BRACS_752', 'BRACS_753', 'BRACS_754', 'BRACS_755', 'BRACS_756', 'BRACS_757', 'BRACS_758', 'BRACS_759', 'BRACS_760', 'BRACS_761', 'BRACS_762', 'BRACS_773', 'BRACS_1228', 'BRACS_1231', 'BRACS_1232', 'BRACS_1233', 'BRACS_1235', 'BRACS_1238', 'BRACS_1239', 'BRACS_1240', 'BRACS_1241', 'BRACS_1242', 'BRACS_1244', 'BRACS_1247', 'BRACS_1248', 'BRACS_1249', 'BRACS_1250', 'BRACS_1251', 'BRACS_1253', 'BRACS_1255', 'BRACS_1257', 'BRACS_1259', 'BRACS_1261', 'BRACS_1263', 'BRACS_1265', 'BRACS_1267', 'BRACS_1269', 'BRACS_1270', 'BRACS_1271', 'BRACS_1272', 'BRACS_1273', 'BRACS_1274', 'BRACS_1275', 'BRACS_1276', '

In [5]:
torch.load("../../../../../../../../mnt/nfs03-R6/CAMELYON16/slides_cls/feats_UNI/pt_files/normal_009.pth").shape

torch.Size([1976, 1024])

In [5]:
# tiles: 217

cmd:
python create_tiles.py --images-dir ../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/train/Group_AT/Type_FEA --save-dir ../../../../../../../../mnt/nfs03-R6/BRACS/gigapath_tiles/train/Group_AT/Type_FEA

Done:
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/train/Group_MT/Type_DCIS: 40 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/train/Group_MT/Type_IC: 100 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/train/Group_AT/Type_FEA: 24 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/train/Group_AT/Type_ADH: 28 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/train/Group_BT/Type_N: 27 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/train/Group_BT/Type_PB: 120 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/train/Group_BT/Type_UDH: 56 files (error with: 1791)
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/val/Group_MT/Type_DCIS: 9 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/val/Group_MT/Type_IC: 12 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/val/Group_AT/Type_FEA: 6 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/val/Group_AT/Type_ADH: 8 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/val/Group_BT/Type_N: 10 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/val/Group_BT/Type_PB: 11 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/val/Group_BT/Type_UDH: 9 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/test/Group_MT/Type_DCIS: 12 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/test/Group_MT/Type_IC: 20 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/test/Group_AT/Type_FEA: 11 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/test/Group_AT/Type_ADH: 12 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/test/Group_BT/Type_N: 7 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/test/Group_BT/Type_PB: 16 files
../../../../../../../../mnt/nfs03-R6/BRACS/BRACS_WSI/test/Group_BT/Type_UDH: 9 files

